bok_choy/a11y/a11y_audit.py

```diff
@@ -0,0 +1,267 @@
+"""
+Interface for running accessibility audits on a PageObject.
+"""
+import json
+import os
+import requests
+
+from abc import abstractmethod, abstractproperty, ABCMeta
+from textwrap import dedent
+
+
+class AccessibilityError(Exception):
+    """
+    The page violates one or more accessibility rules.
+    """
+    pass
+
+
+class A11yAuditConfig(object):
+    """
+    The `A11yAuditConfig` object defines the options available in an
+    accessibility ruleset.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, *args, **kwargs):
+        super(A11yAuditConfig, self).__init__(*args, **kwargs)
+        self.rules_file = None
+
+    def set_rules_file(self, path=None):
+        """
+        Sets `self.rules_file` to the passed file.
+
+        Args:
+            A filepath where the JavaScript for the ruleset can be found.
+
+        This is intended to be used in the case of using an extended or
+        modified version of the ruleset. The interface and response
+        format are expected to be unmodified.
+        """
+        if path:
+            self.rules_file = os.path.abspath(path)
+
+    @abstractmethod
+    def set_rules(self, rules, ignore=False, include_type="rule"):
+        """
+        Overrides the default list of rules to be run.
+
+        Raises:
+
+            `NotImplementedError` if this isn't overwritten in the ruleset
+                specific implementation.
+        """
+        raise NotImplementedError(
+            "The ability to specify rules has not been implemented."
+        )
+
+    @abstractmethod
+    def set_scope(self, include=None, exclude=None):
+        """
+        Overrides the default scope (part of the DOM) to inspect.
+
+        Raises:
+
+            `NotImplementedError` if this isn't overwritten in the ruleset
+                specific implementation.
+        """
+        raise NotImplementedError(
+            "The ability to specify scope has not been implemented."
+        )
+
+
+class A11yAudit(object):
+    """
+    Allows auditing of a page for accessibility issues.
+
+    The ruleset to use can be specified by the environment variable
+    `BOKCHOY_A11Y_RULESET`. Currently, there are two ruleset implemented:
+
+    `axe_core`:
+
+        * Ruleset class: AxeCoreAudit
+        * Ruleset config: AxeCoreAuditConfig
+        * This is default ruleset.
+
+    `google_axs`:
+
+        * Ruleset class: AxsAudit
+        * Ruleset config: AxsAuditConfig
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, browser, url, config=None, *args, **kwargs):
+        """
+        Sets ruleset to be used.
+
+        Args:
+            browser: A phantomjs browser instance
+            url: URL of the page to test
+            config: (optional) A11yAuditConfig or subclass of A11yAuditConfig
+        """
+        super(A11yAudit, self).__init__(*args, **kwargs)
+        self.url = url
+        self.browser = browser
+        self.config = config or self.default_config
```

Comment:
@jzoldak `self.default_config` is a property of a `A11yAudit` (see [line 202](https://github.com/edx/bok-choy/pull/109/files#diff-8650032c466196a84557f7f5368250d5R222))